llvm/lib/CodeGen/MachineScheduler.cpp

```diff
@@ -4266,6 +4266,12 @@ unsigned ResourceSegments::getFirstAvailableAt(
   assert(std::is_sorted(std::begin(_Intervals), std::end(_Intervals),
                         sortIntervals) &&
          "Cannot execute on an un-sorted set of intervals.");
+
+  // Zero resource usage is allowed by TargetSchedule.td but we do not construct
+  // a ResourceSegment interval for that situation.
```

Comment:
> It should return CurrCycle since it is as if the for loop below hit the continue statement every time since nothing intersects with the empty interval.

I see, so this is an optimisation of the algorithm, not a functional change. It would be best not to add such optimisations (more code, more bugs!), unless there is a significant slow down in compile time.

If you want this optimisation, I would rephrase the comment as 

```suggestion
  // Empty resource intervals (`AcquireAtCycle == ReleaseAtCycle`)
  // are not stored in a ResourcesSegment.
  // However, zero resource usage is allowed by TargetSchedule.td.
  // If we deal with an empty interval, we can just return CurrCycle since
  // it is as if the for loop below hit the continue statement every time, 
  // since nothing intersects with the empty interval.
```

Reply:
This is not an optimization. It was made because intersect will crash when considering the empty interval since intervals are supposed to be [a, b), not (a, b)